webhooks/clusterclass.go

```diff
@@ -123,8 +140,126 @@ func (webhook *ClusterClass) validate(old, new *clusterv1.ClusterClass) error {
+
 	if len(allErrs) > 0 {
 		return apierrors.NewInvalid(clusterv1.GroupVersion.WithKind("ClusterClass").GroupKind(), new.Name, allErrs)
 	}
 	return nil
 }
+
+func (webhook *ClusterClass) validateLivingMachineDeploymentClassesNotRemoved(old, new *clusterv1.ClusterClass) field.ErrorList {
+	var allErrs field.ErrorList
+	if old == nil {
+		return allErrs
+	}
+	missingClasses := webhook.removedMachineClasses(old, new)
+
+	// If no classes have been removed return early as no further checks are needed.
+	if len(missingClasses) == 0 {
+		return allErrs
+	}
+
+	// TODO: The first part of this method will need to be broken out to properly account for changes when also validating variables.
+	// Retrieve all clusters using the ClusterClass.
+	clusters, errs := webhook.clustersUsingClusterClass(old)
+	if len(errs) > 0 {
+		allErrs = append(allErrs, errs...)
+		return allErrs
+	}
+
+	// Retrieve all the MachineDeployments in the clusterClass namespace using a managed topology.
+	machineDeployments, errs := webhook.topologyManagedMachineDeployments(new.Namespace)
+	if len(errs) > 0 {
+		allErrs = append(allErrs, errs...)
+		return allErrs
+	}
+
+	// Create a set of records of machineDeploymentClass information where the name is a fully qualified name for the
+	// MachineDeployment topology in the form cluster.Name/machineDeploymentTopology.Name and links to the machineDeploymentClass Name.
+	mdcRecords := map[string]string{}
+	for _, c := range clusters {
+		for _, machineDeploymentClass := range c.Spec.Topology.Workers.MachineDeployments {
+			if missingClasses.Has(machineDeploymentClass.Class) {
+				mdcRecords[fmt.Sprintf("%v/%v", c.Name, machineDeploymentClass.Name)] = machineDeploymentClass.Class
+			}
+		}
+	}
+
+	// For each machineDeployment using a managed topology return an error if it is using a class that has been removed in the ClusterClass change.
+	for _, md := range machineDeployments {
+		mdName := fmt.Sprintf("%v/%v", md.Labels[clusterv1.ClusterLabelName], md.Labels[clusterv1.ClusterTopologyMachineDeploymentLabelName])
+		if _, ok := mdcRecords[mdName]; ok {
+			allErrs = append(allErrs, field.Forbidden(field.NewPath(""), fmt.Sprintf("MachineDeploymentClass %v is in use in MachineDeployment %v in Cluster %v. ClusterClass %v modification not allowed",
+				mdcRecords[mdName], md.Labels[clusterv1.ClusterTopologyMachineDeploymentLabelName], md.Labels[clusterv1.ClusterLabelName], old.Name),
+			))
+		}
+	}
+
+	return allErrs
+}
+
+func (webhook *ClusterClass) topologyManagedMachineDeployments(namespace string) ([]clusterv1.MachineDeployment, field.ErrorList) {
+	var allErrs field.ErrorList
+
+	// List all the MachineDeployments in the current cluster using a managed topology.
+	machineDeployments := &clusterv1.MachineDeploymentList{}
+	err := webhook.Client.List(context.Background(), machineDeployments,
+		client.MatchingLabels{
+			clusterv1.ClusterTopologyOwnedLabel: "",
+		},
+		// TODO: Should this be namespaced in future?
```

Comment:
Seems to be already namespaced (or maybe you mean that differently)

Reply:
Yeah - there's an issue to modify the way namespacing for clusterclass works #5673. I just wanted to call out the parts of this code that will be impacted.